### ISSUE TITLE
LSTMs with Sequences for outputs

### ISSUE THREAD

**Issue by AntreasAntoniou** [fchollet/keras#1110]
So I've been playing around with regenerative modeling using RNNs and I've been wondering the following:

Assuming an input of size (1000, 25, 15) where 1000 is the samples, 25 is the sequences per sample and 15 are the features per sequence. I am predicting the next sequence as output and it works well. Now if I wanted to predict further in the future I could feed the next prediction as an input and go on for as long as I want. My question is, could I train a model based on n_number of sequences as y_train? E.g instead of getting for outputs (1000, 1, 15), to get as output (1000, 25, 15) therefore training a model to predict a number of future sequences. I have already hacked something together that does that but not in the nicest of ways. Basically I train 15 models, one for each feature that all predict the next 25 outputs for that value and then I stitch them together. Does keras support a way of doing this without having to train a gazillion models each time? Thanks for your time.

**Comment by rpinsler** [comment]
Sounds like you want to use `return_sequences=True`. I don't know how if the provided loss functions can handle sequences automatically, otherwise you'd have to implement your own cost function that does that.

**Comment by AntreasAntoniou** [comment]
Yeah thanks. That was very helpful. I found a way to do something very similar to what I wanted to do. I'll try it out.

**Comment by thenomemac** [comment]
Please comment if you're successful. As have an identical use case and have been trying to debug how to set this up correctly.

**Comment by swayson** [comment]
I am also curious as to how to implement this.

**Comment by AntreasAntoniou** [comment]
Hey guys, 

I did the code for this, and it worked very well. So this is how to do it.

Given 1000, 50, 24 input and 1000, 50, 24 outputs 
output_features=X.shape[2]
model = Sequential()
        model.add(LSTM(neuron_units, return_sequences=True, input_shape=(self.X_full.shape[1], self.X_full.shape[2])))
        model.add(Dropout(0.2))
        model.add(TimeDistributedDense(output_dim=output_features))
        print("Output Units "+str(output_features))
        model.add(Activation('tanh'))
        model.compile(loss='mean_squared_error', optimizer='rmsprop')
        return model

**Comment by thenomemac** [comment]
One question on your example. Why did you output 24 dim for your features. Are your 24i nput features lags of the variable your trying to predict and your 24 output features the next 24 observations of the variable or are you actually trying to predict 24 different variables simultaneously 50 periods into the future? Just curious as to what exactly you were able to be successful with?

**Comment by AntreasAntoniou** [comment]
My sample changed since last time, so 1000 is the number of samples, 50 is the number of sequences per sample and 24 is the number of features per sequence. I feed that in and get predictions on the next 50 sequences for all 24 of their features. The performance is really really good. It has managed to learn very complex signals and even managed to predict features that are otherwise set by a human(control signals).

**Comment by thenomemac** [comment]
Really curious what example you have with 24 features that need predictions. I can think of lots of examples where your input is (1000,50,24) and output is (1000,24,1) for example input features are 24 current economic indicators and output singular feature is next 50 gdp values. Are your predicting 24 separate time series that are not independent like server logs or stocks in the Dow Jones? Or do you really only care about the prediction for 1 of your 24 features are the rest are just correlated with the one you care about but are primaryily just used for inputs?

**Comment by AntreasAntoniou** [comment]
I get what your point is. My example is basically the behavior of a system/machine and everything is connected to each other. In more general cases, you may use it to predict the next frame from the previous frame which would obsiously be a very large amount of features. Also in generative modeling you feed all the outputs back to the inputs to hallucinate what comes next. So yeah, that's what I did.

**Comment by jfsantos** [comment]
@AntreasAntoniou Please close this, since it's not a bug and your problem was solved. Questions should be sent to the [mailing list](https://groups.google.com/forum/#!forum/keras-users) instead of the issue tracker.

**Comment by AntreasAntoniou** [comment]
@jfsantos For some reason I don't have access to the close button.

**Comment by ersinyar** [comment]
@AntreasAntoniou What did you use as the number of LSTM blocks in the hidden layer? Also can you share your whole code please?

**Comment by yangfantrinity** [comment]
@AntreasAntoniou , I am working on a similar problem. The main difference is that I am trying to use an input of (1000, 96, 1) to get the output of (1000, 8, 1), which means the number of time steps in input and output is different. In other words, I am trying to use the previous 96-time steps value to predict the next 8-time steps value. Is it possible to use the same approach to achieve this? I tried to run but it gives me the error:
**Error when checking model target: expected activation_1 to have shape (None, 96, 1) but got array with shape (1000, 9, 1) .** 

Alternatively, I have another way to formulate this problem. I view the previous 96-time steps as one time step, and I use 96 as the number of features. Hence my input looks like (1000, 1, 96) and the output looks like (1000, 1, 8). I am not so sure about this, though.
